Refactored header files structure to improve consistency with lib/

andreamattavelli opened the pull request:
Improved layout of include/  to match the layout of lib/ as discussed in #324

delcypher commented on include/klee/Core/ExecutionState.h:
```diff
@@ -10,18 +10,16 @@
 #include <set>
 #include <vector>
 
+#include "klee/Support/ADT/TreeStream.h"
```
Please sort includes alphabetically. You can have clang-format do this for you.

delcypher commented on include/klee/Expr/Expr.h:
```diff
@@ -23,6 +20,8 @@
 #include <vector>
 #include <map>
+#include "klee/Support/Bits.h"
+#include "Ref.h"
```
For consistency that could be `#include "klee/Expr/Ref.h"`. I'm not sure though. What's our policy of library (module) local includes?

andreamattavelli replied:
Good question, according to your link should be:

Main Module Header
Local/Private Headers
LLVM project/subproject headers (clang/..., lldb/..., llvm/..., etc)
System #includes

So, if I've understood correctly, it should look like this:
```
#include "klee/Support/Bits.h"

#include "Ref.h"

#include "llvm/ADT/APInt.h"
#include "llvm/ADT/APFloat.h"
#include "llvm/ADT/DenseSet.h"
#include "llvm/ADT/SmallVector.h"
#include "llvm/Support/raw_ostream.h"

#include <sstream>
#include <set>
#include <vector>
#include <map>
```

Correct?

delcypher replied:
Almost but KLEE is a LLVM sub project so this should be

```
#include "Ref.h"

#include "klee/Support/Bits.h"
#include "llvm/ADT/APInt.h"
#include "llvm/ADT/APFloat.h"
#include "llvm/ADT/DenseSet.h"
#include "llvm/ADT/SmallVector.h"
#include "llvm/Support/raw_ostream.h"

#include <sstream>
#include <set>
#include <vector>
```

andreamattavelli replied:
Ok, so I should treat KLEE as LLVM subproject. Got it (hopefully)

delcypher commented on include/klee/Solver/Solver.h:
```diff
@@ -10,11 +10,11 @@
-
 #include <vector>
 
+#include "../Basic/CommandLine.h" // FIXME: This is just for CoreSolverType
```
No relative includes! Do `#include "klee/Basic/CommandLine.h"`.

andreamattavelli replied:
I've tried my best to remove all the relative includes already present in KLEE, unfortunately I missed some. Good catch!

delcypher commented on include/klee/Support/FloatEvaluation.h:
```diff
@@ -12,14 +12,14 @@
 
 #include <cassert>
 
+#include "Bits.h"     //bits64::truncateToNBits
```
Again for consistency should probably be `#include "klee/Support/Bits.h"`.

delcypher replied:
Disregard this comment. Library local includes are fine.

andreamattavelli replied:
Don't worry, I am applying LLVM guidelines

delcypher commented on include/klee/Support/IntEvaluation.h:
```diff
@@ -10,7 +10,7 @@
 #define KLEE_UTIL_INTEVALUATION_H
 
-#include "klee/util/Bits.h"
+#include "Bits.h"
```
Again for consistency should probably be `#include "klee/Support/Bits.h"`.

delcypher replied:
Disregard this comment. Library local includes are fine.

delcypher commented on include/klee/Support/PrintContext.h:
```diff
@@ -10,11 +10,11 @@
 #include <sstream>
 #include <string>
 #include <stack>
+#include "klee/Expr/Expr.h"
```
Quoted includes should come before `<>` style includes. See http://llvm.org/docs/CodingStandards.html#include-style

delcypher commented on include/klee/Support/QueryLog.h:
```diff
@@ -10,9 +10,8 @@
-#include "klee/Expr.h"
-
 #include <vector>
+#include "klee/Expr/Expr.h"
```
Again. Include order is wrong ( http://llvm.org/docs/CodingStandards.html#include-style )

delcypher commented on lib/Basic/CmdLineOptions.cpp:
```diff
@@ -12,7 +12,7 @@
  */
 
-#include "klee/CommandLine.h"
+#include "../../include/klee/Basic/CommandLine.h"
```
Ah! No relative includes! This should be `#include "klee/Basic/CommandLine.h"`

delcypher commented on lib/Basic/ConstructSolverChain.cpp:
```diff
@@ -10,11 +10,12 @@
 #include "llvm/Support/raw_ostream.h"
 
+#include "../../include/klee/Basic/CommandLine.h"
+#include "../../include/klee/Basic/Common.h"
```
Oh no! More relative includes! These should be `#include "klee/Basic/CommandLine.h` and `#include "klee/Basic/Common.h"`.

delcypher commented on lib/Core/CallPathManager.h:
```diff
@@ -10,10 +10,9 @@
-
 #include <map>
 #include <vector>
+#include "klee/Basic/Statistics.h"
```
Wrong include order quoted includes should come before system includes.

delcypher commented on lib/Core/Executor.cpp:
```diff
@@ -8,6 +8,29 @@
 
 #include "Executor.h"
+
+#include "../../include/klee/Basic/CommandLine.h"
```
Again. No relative includes!

delcypher commented on lib/Core/ImpliedValue.cpp:
```diff
@@ -10,17 +10,15 @@
+#include "klee/Expr/Expr.h"
+#include "klee/Expr/ExprUtil.h"
+#include "klee/Solver/Solver.h"
+#include "klee/Support/IntEvaluation.h"
```
Don't drop the FIXME unless this is actually fixed.

delcypher commented on lib/Core/ImpliedValue.h:
```diff
@@ -10,9 +10,8 @@
-#include "klee/Expr.h"
-
 #include <vector>
+#include "klee/Expr/Expr.h"
```
Wrong include order. Quoted includes come before system includes.

delcypher commented on lib/Core/Memory.cpp:
```diff
@@ -9,13 +9,12 @@
+#include "klee/Expr/Expr.h"
+#include "klee/Solver/Solver.h"
+#include "klee/Support/ErrorHandling.h"
 #include "Context.h"
```
Local header `Context.h` should come before these if are allowing library local includes (which this file already has).

delcypher commented on lib/Core/MemoryManager.cpp:
```diff
@@ -11,13 +11,13 @@
 #include <sys/mman.h>
+
+#include "klee/Expr/Expr.h"
+#include "klee/Support/ErrorHandling.h"
```
Should come before system includes.

delcypher commented on lib/Core/Searcher.cpp:
```diff
@@ -9,21 +9,21 @@
+#include "klee/Support/ErrorHandling.h"
+#include "klee/Support/ModuleUtil.h"
+#include "klee/Support/Time.h"
 #include "CoreStats.h"
```
Library local includes should come before the other library headers.

delcypher commented on lib/Core/SpecialFunctionHandler.cpp:
```diff
@@ -9,19 +9,18 @@
-#include "klee/Internal/Module/KModule.h"
-#include "klee/Internal/Support/Debug.h"
-#include "klee/Internal/Support/ErrorHandling.h"
+#include "../../include/klee/Basic/CommandLine.h"
```
No relative includes!

delcypher commented on lib/Expr/ExprPPrinter.cpp:
```diff
@@ -7,16 +7,15 @@
 
 #include <map>
 #include <vector>
+#include "klee/Expr/Constraints.h"
```
Include should come before system includes.

delcypher commented on lib/Expr/Lexer.cpp:
```diff
@@ -7,7 +7,7 @@
 //===----------------------------------------------------------------------===//
 
-#include "expr/Lexer.h"
+#include "../../include/klee/Expr/Parser/Lexer.h"
```
Ah! No relative includes!

delcypher commented on lib/Expr/Parser.cpp:
```diff
@@ -26,6 +18,13 @@
 #include <map>
 #include <cstring>
 
+#include "../../include/klee/Expr/Parser/Lexer.h"
```
No relative includes!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

delcypher commented on lib/SMT/SMTParser.cpp:
```diff
@@ -9,17 +9,17 @@
 #include <cassert>
 #include <stack>
 
+#include "../../include/klee/Expr/expr/Parser.h"
```
No relative includes and the `klee/*` includes should come before the system ones.

delcypher commented on lib/SMT/SMTParser.h:
```diff
@@ -11,12 +11,12 @@
 #include <stack>
 #include <string>
 
+#include "../../include/klee/Expr/Parser/Parser.h"
```
No relative includes!

delcypher commented on lib/Solver/CachingSolver.cpp:
```diff
@@ -8,16 +8,16 @@
+#include "klee/Solver/SolverStats.h"
 
 #include <ciso646>
+
```
`klee/*` includes should come before system includes.

delcypher commented on lib/Solver/ConstantDivision.cpp:
```diff
@@ -9,10 +9,9 @@
-
 #include <algorithm>
 #include <cassert>
+#include "klee/Support/Bits.h"
```
`klee/*` include should be before system includes.

delcypher commented on lib/Solver/CoreSolver.cpp:
```diff
@@ -7,12 +7,13 @@
 #include "llvm/Support/raw_ostream.h"
 #include <string>
 
+#include "../../include/klee/Basic/CommandLine.h"
```
No relative includes!!!!!!!!

delcypher commented on lib/Solver/STPBuilder.cpp:
```diff
@@ -10,10 +10,10 @@
-#include "klee/SolverStats.h"
+#include "klee/Expr/Expr.h"
+#include "klee/Solver/Solver.h"
+#include "klee/Support/Bits.h"
```
should be alphabetically ordered.

delcypher commented on lib/Solver/STPBuilder.h:
```diff
@@ -10,12 +10,13 @@
 
 #include <vector>
 
+#include "klee/Expr/ArrayExprHash.h"
```
`klee/*` include should come before system include.

delcypher commented on lib/Solver/ValidatingSolver.cpp:
```diff
@@ -6,10 +6,10 @@
-#include "klee/Solver.h"
-#include "klee/SolverImpl.h"
 #include <vector>
+#include "klee/Expr/Constraints.h"
```
`klee/*` include should come before system include.

delcypher commented on lib/Solver/Z3Builder.h:
```diff
@@ -10,11 +10,12 @@
 #include "klee/Config/config.h"
 #include <z3.h>
 
+#include "klee/Expr/ArrayExprHash.h"
```
`klee/*` include should come before system include.

delcypher commented on lib/Support/Time.cpp:
```diff
@@ -7,9 +7,9 @@
-#include "klee/Internal/System/Time.h"
+#include "klee/Support/Time.h"
 
+#include "klee/Config/Version.h"
```
should be alphabetically ordered.

delcypher commented on tools/kleaver/main.cpp:
```diff
@@ -36,6 +20,21 @@
 
 #include "llvm/Support/Signals.h"
 
+#include "../../include/klee/Basic/CommandLine.h"
```
AH! No relative includes!

delcypher commented on tools/klee-replay/klee-replay.c:
```diff
@@ -24,6 +23,7 @@
 #include <unistd.h>
 #include <sys/signal.h>
 #include <sys/wait.h>
+#include "klee/Support/ADT/KTest.h"
```
`klee/*` includes should be before system includes.

delcypher commented:
@andreamattavelli Some high level comments for style problems I saw repeated.

* You should follow http://llvm.org/docs/CodingStandards.html#include-style
* You need to fix the Z3 and MetaSMT builds
* Never ever use relative includes (i.e `#include "../Basic/CommandLine.h"`), use `#include "klee/Basic/CommandLine.h`).
* Make sure KLEE header files (e.g. `klee/Expr/Expr.h`) come before system includes (e.g. `<vector>`).
* Library local includes should come before `klee/*` includes.

ccadar commented:
I'd like to take a look at this before merging it, to make sure I can still find my way through the code afterwards :)

delcypher commented:
@andreamattavelli When you're done fixing the issues and have pushed a new change let me know and I'll review again.

andreamattavelli commented:
@delcypher First of all, thanks for reviewing the PR. Most of the issues were introduced by Eclipse refactoring operations.

Some changes:
* Applied the LLVM code style guide wherever possible
* Fixed the Z3 and MetaSMT builds
* Removed all relative paths

Still, I might have missed something. Thanks for checking.

delcypher commented on include/klee/Core/ExecutionState.h:
```diff
@@ -10,13 +10,12 @@
-#include "klee/Internal/ADT/TreeStream.h"
-
-// FIXME: We do not want to be exposing these? :(
 #include "../../lib/Core/AddressSpace.h"
```
Don't drop the FIXME if you don't actually fix the problem. So either put the FIXME back on move `AddressSpace.h` into `include/klee/Core` so it can be included as `#include "klee/Core/AddressSpace.h"`

delcypher commented on include/klee/Support/CompressionStream.h:
```diff
@@ -7,8 +7,8 @@
+#define INCLUDE_KLEE_SUPPORT_COMPRESSIONSTREAM_H_
 
 #include "llvm/Support/raw_ostream.h"
 #include "zlib.h"
```
That `#include "zlib.h"` probably ought to be `#include <zlib.h>`

delcypher commented on include/klee/Support/FloatEvaluation.h:
```diff
@@ -12,9 +12,9 @@
-#include "IntEvaluation.h" //ints::sext
+#include "Bits.h" //bits64::truncateToNBits
 
+#include "klee/Support/IntEvaluation.h" //ints::sext
```
This is part of `klee/Support` so `#include "IntEvaluation.h"` should be sufficient. It's weird to do `#include "Bits.h"` but then do `#include "klee/Support/IntEvaluation.h"`.

delcypher commented on lib/Core/ExecutionState.cpp:
```diff
@@ -7,16 +7,15 @@
+#include "Memory.h"
 
-#include "klee/Expr.h"
+#include "klee/Core/ExecutionState.h"
```
Because this implements `ExecutionState` shouldn't `klee/Core/ExecutionState.h` be that first include?

delcypher commented on lib/Core/Executor.cpp:
```diff
@@ -10,6 +10,8 @@
 #include "Executor.h"
 #include "Context.h"
 #include "CoreStats.h"
+#include "Executor.h"
```
`#include "Executor.h"` is already there. Why are you adding it again?

delcypher commented on lib/Core/Memory.cpp:
```diff
@@ -7,17 +7,16 @@
-
 #include "ObjectHolder.h"
 #include "MemoryManager.h"
+#include "Memory.h"
```
This file Implements `Memory` so shouldn't `Memory.h` be the first include?

delcypher commented on lib/Core/MemoryManager.cpp:
```diff
@@ -11,13 +11,13 @@
 #include "Memory.h"
 #include "MemoryManager.h"
```
Again this is the implementation for `MemoryManager` so `MemoryManager.h` should be the first include.

delcypher commented on lib/Core/Searcher.cpp:
```diff
@@ -7,23 +7,23 @@
 #include "CoreStats.h"
 #include "Executor.h"
 #include "PTree.h"
+#include "Searcher.h"
```
This file implements `Searcher` so `Searcher.h` should be the first include.

delcypher commented on lib/Core/SpecialFunctionHandler.cpp:
```diff
@@ -7,21 +7,18 @@
+#include "Executor.h"
 #include "Memory.h"
+#include "MemoryManager.h"
 #include "SpecialFunctionHandler.h"
```
This implements `SpecialFunctionHandler` so `SpecialFunctionHandler.h` should come first.

delcypher commented on lib/Core/StatsTracker.cpp:
```diff
@@ -7,26 +7,25 @@
 #include "CoreStats.h"
 #include "Executor.h"
 #include "MemoryManager.h"
+#include "StatsTracker.h"
```
This implements `StatsTracker` so `StatsTracker.h` should come first.

delcypher commented on lib/Core/TimingSolver.cpp:
```diff
@@ -7,16 +7,14 @@
 //===----------------------------------------------------------------------===//
 
+#include "CoreStats.h"
 #include "TimingSolver.h"
```
This implements `TimingSolver` so `TimingSolver.h` should come first.

delcypher commented on lib/Core/UserSearcher.cpp:
```diff
@@ -7,12 +7,11 @@
-#include "Searcher.h"
 #include "Executor.h"
+#include "Searcher.h"
+#include "UserSearcher.h"
```
This implements `UserSearcher` so `UserSearcher.h` should come first.

delcypher commented on lib/Expr/Constraints.cpp:
```diff
@@ -7,17 +7,16 @@
-
-#include "klee/util/ExprPPrinter.h"
-#include "klee/util/ExprVisitor.h"
+#include "klee/Expr/Constraints.h"
```
This file implements `Constraints` so `klee/Expr/Constraints.h` should come first.

delcypher commented on lib/Expr/Expr.cpp:
```diff
@@ -7,19 +7,18 @@
 
-#include "klee/Expr.h"
 #include "klee/Config/Version.h"
+#include "klee/Expr/Expr.h"
```
The implements `Expr` so `klee/Expr/Expr.h` should come first.

delcypher commented on lib/Expr/ExprUtil.cpp:
```diff
@@ -7,12 +7,10 @@
+#include "klee/Expr/Expr.h"
+#include "klee/Expr/ExprHashMap.h"
+#include "klee/Expr/ExprVisitor.h"
+#include "klee/Expr/ExprUtil.h"
```
This implements `ExprUtil` so `klee/Expr/ExprUtil.h` should come first.

MartinNowack commented:
This patch has been mostly superseded by #1143 and #1122. The related discussion is still in #324 . @andreamattavelli Thanks a lot for this huge effort in the first place.
